File: forge/src/main/java/org/valkyrienskies/clockwork/forge/content/contraptions/curiosities/tools/gravitron/tool/GrabTool.java
package org.valkyrienskies.clockwork.forge.content.contraptions.curiosities.tools.gravitron.tool;


import com.simibubi.create.AllPackets;
import com.simibubi.create.content.schematics.client.tools.ToolType;
import com.simibubi.create.content.schematics.packet.SchematicPlacePacket;
import net.minecraft.core.BlockPos;
import net.minecraft.nbt.Tag;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.level.Level;
import net.minecraft.world.phys.Vec3;
import net.minecraftforge.event.entity.living.LivingEvent;
import org.joml.Quaterniond;
import org.joml.Vector2d;
import org.joml.Vector3d;
import org.joml.Vector3dc;
import org.valkyrienskies.clockwork.ClockworkItems;
import org.valkyrienskies.clockwork.ClockworkPackets;
import org.valkyrienskies.clockwork.content.curiosities.tools.gravitron.GravitronForceInducer;
import org.valkyrienskies.clockwork.content.curiosities.tools.gravitron.GravitronForceInducerData;
import org.valkyrienskies.clockwork.content.curiosities.tools.gravitron.GravitronItem;
import org.valkyrienskies.clockwork.forge.content.contraptions.curiosities.tools.gravitron.GravitronGrabPacket;
import org.valkyrienskies.clockwork.util.ClockworkUtils;
import org.valkyrienskies.core.api.ships.LoadedServerShip;
import org.valkyrienskies.mod.common.VSGameUtilsKt;
import org.valkyrienskies.mod.common.util.VectorConversionsMCKt;

import static java.lang.Math.toRadians;

public class GrabTool extends GravitronToolBase {

    public static void tick(LivingEvent.LivingTickEvent event) {
        var entity = event.getEntity();
        if (entity instanceof Player player && player.level() instanceof ServerLevel serverLevel) {
            var s = getState(player);
            var graviton = player.getMainHandItem();
            //System.out.println(s.getShipID() + " : " + s.getGrabbing());
            if (true) {
                if (!s.getShouldDrop() && graviton.is(ClockworkItems.GRAVITRON.asItem())) {
                    updateShip(s, serverLevel, entity);
                } else {
                    dropShip(s, serverLevel);
                }
                //MAIN LEVEL

                if (s.getGrabCD() != null && s.getGrabCD() > 0) {
                    s.setGrabCD(s.getGrabCD() - 1);
                }
                if (graviton.hasTag() && graviton.getTag().contains("GrabbedPosInShip") && !player.getCooldowns().isOnCooldown(graviton.getItem())) {
                    s.setGrabbing(true);
                    var tag = graviton.getTag();

                    var clickLocation = ClockworkUtils.readVec3(tag.getList("GrabbedPosInShip", Tag.TAG_DOUBLE));
                    var id = tag.getLong("ShipId");

                    var ship = VSGameUtilsKt.getShipObjectWorld(serverLevel).getLoadedShips().getById(id);
                    if (ship != null) {
                        System.out.println(ship);
                        var transformedPos = ship.getWorldToShip().transformPosition(VectorConversionsMCKt.toJOML(clickLocation), new Vector3d());
                        grabShip(s, player, ship, transformedPos);
                        graviton.removeTagKey("ShipId");
                        graviton.removeTagKey("GrabbedPosInShip");
                    }
                }
            } else {

            }
        }

    }

    private static void dropShip(GravitronItem.Companion.GravitronState s, ServerLevel level) {
        var grabbedShipId = s.getShipID();
        if (grabbedShipId != null) {
            var loadedShip = VSGameUtilsKt.getShipObjectWorld(level).getLoadedShips().getById(grabbedShipId);
            if (loadedShip != null) {
                var gravitronForceInducer = GravitronForceInducer.Companion.getOrCreate(loadedShip);
                gravitronForceInducer.setData(null);
            }
        }

        s.setGrabbing(false);
        s.setShipID(null);
        s.setShouldDrop(false);
    }

    private static void updateShip(GravitronItem.Companion.GravitronState s , ServerLevel level, Entity entity) {
        if (s.getGrabbing()) {
            var shipId = s.getShipID();
            if (shipId != null) {
                var shipUnloaded = VSGameUtilsKt.getShipObjectWorld(level).getAllShips().getById(shipId);
                var ship = VSGameUtilsKt.getShipObjectWorld(level).getLoadedShips().getById(shipId);
                if (ship != null && s.getPlayerGrabbedRotation() != null && s.getShipGrabbedDistance() != null && s.getShipGrabbedPos() != null && s.getHeldBlockPos() != null) {
                    // Update Rot Values
                    var playerCurrentRotation = new Vector2d(entity.getXRot(), entity.getYRot());
                    var origPlayerRot = playerRotToQuaternion(s.getPlayerGrabbedRotation().x(), s.getPlayerGrabbedRotation().y()).normalize();
                    var newPlayerRot = playerRotToQuaternion(playerCurrentRotation.x(), playerCurrentRotation.y()).normalize();
                    var deltaPlayerRot = newPlayerRot.mul(origPlayerRot.conjugate(new Quaterniond()), new Quaterniond());
                    var rotation = deltaPlayerRot.mul(s.getShipGrabbedRot(), new Quaterniond()).normalize();

                    // Update Pos Values
                    var lookDif = VectorConversionsMCKt.toJOML(entity.getLookAngle()).normalize().mul(s.getShipGrabbedDistance());
                    s.setHeldBlockPos(VectorConversionsMCKt.toJOML(entity.getEyePosition()).add(lookDif));
                    var location = new Vector3d(s.getShipGrabbedPos());
                    var position = new Vector3d(s.getHeldBlockPos());

                    var gravitronForceInducer = GravitronForceInducer.Companion.getOrCreate(ship);
                    var newData = new GravitronForceInducerData(position, rotation, location);
                    gravitronForceInducer.setData(newData);
                } else if (shipUnloaded == null) {
                    dropShip(s, level);
                }
            }
        }
    }

    private static Quaterniond playerRotToQuaternion(double pitch, double yaw) {
        return new Quaterniond().rotateY(toRadians(-yaw)).rotateX(toRadians(pitch));
    }

    @Override
    public boolean handleRightClick() {
        updateTargetPos();

        ClockworkPackets.sendToServer(new GravitronGrabPacket(clickedPos, clickedLocation));

        return false;
    }

    @Override
    public boolean handleMouseWheel(double delta) {
        return false;
    }

    public static boolean tryGrabShip(ServerLevel level, Player player, BlockPos clickedPos, Vec3 clickLocation) {
        int chunkX = clickedPos.getX() >> 4;
        int chunkZ = clickedPos.getZ() >> 4;
        System.out.println("PreGrab0: " + clickedPos + " : " + clickLocation);
        LoadedServerShip ship = VSGameUtilsKt.getShipObjectWorld(level).getLoadedShips().getByChunkPos(chunkX, chunkZ, VSGameUtilsKt.getDimensionId(level));
        Vector3dc grabPosInShip = VectorConversionsMCKt.toJOML(clickLocation);
        Vector3d grabPosInWorld = new Vector3d(grabPosInShip);
        var s = getState(player);
        System.out.println("PreGrab: " + grabPosInShip + " : " + grabPosInWorld);
        if (VSGameUtilsKt.isBlockInShipyard(level, clickedPos) && ship == null) {
            return false;
        }

        if (ship == null) {
            return false;
        } else {
            ship.getShipToWorld().transformPosition(grabPosInWorld);
        }
        System.out.println("Grab");
        grabShip(s, player, ship, grabPosInShip);
        return true;
    }

    private static void grabShip(GravitronItem.Companion.GravitronState s, Player p, LoadedServerShip ship, Vector3dc grabPosInShip) {
        Vector3d heldPosInWorld = new Vector3d();
        ship.getTransform().getShipToWorld().transformPosition(new Vector3d(grabPosInShip), heldPosInWorld);

        s.setShipID(ship.getId());
        s.setHeldBlockPos(heldPosInWorld);
        s.setPlayerGrabbedRotation(new Vector2d(p.getXRot(), p.getYRot()));
        s.setShipGrabbedPos(new Vector3d(grabPosInShip));
        s.setShipGrabbedRot(ship.getTransform().getShipToWorldRotation());
        s.setShipGrabbedDistance(VectorConversionsMCKt.toJOML(p.getEyePosition()).distance(heldPosInWorld));
        ship.setStatic(false);
    }
}

File: common/src/main/kotlin/org/valkyrienskies/clockwork/content/curiosities/tools/gravitron/tool/GrabTool.kt
package org.valkyrienskies.clockwork.content.curiosities.tools.gravitron.tool

import com.simibubi.create.AllKeys
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.nbt.Tag
import net.minecraft.server.level.ServerLevel
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.player.Player
import net.minecraft.world.phys.Vec3
import org.joml.Quaterniond
import org.joml.Vector2d
import org.joml.Vector3d
import org.joml.Vector3dc
import org.valkyrienskies.clockwork.ClockworkItems
import org.valkyrienskies.clockwork.ClockworkPackets.Companion.sendToServer
import org.valkyrienskies.clockwork.content.curiosities.tools.gravitron.GravitronForceInducer.Companion.getOrCreate
import org.valkyrienskies.clockwork.content.curiosities.tools.gravitron.GravitronForceInducerData
import org.valkyrienskies.clockwork.content.curiosities.tools.gravitron.GravitronGrabPacket
import org.valkyrienskies.clockwork.content.curiosities.tools.gravitron.GravitronItem.Companion.GravitronState
import org.valkyrienskies.clockwork.content.curiosities.tools.gravitron.GravitronItem.Companion.getState
import org.valkyrienskies.clockwork.util.ClockworkUtils.readVec3
import org.valkyrienskies.core.api.ships.LoadedServerShip
import org.valkyrienskies.mod.common.dimensionId
import org.valkyrienskies.mod.common.isBlockInShipyard
import org.valkyrienskies.mod.common.shipObjectWorld
import org.valkyrienskies.mod.common.util.toJOML

class GrabTool : GravitronToolBase() {

    override fun handleRightClick(): Boolean {
        updateTargetPos()
        sendToServer(GravitronGrabPacket(clickedPos, clickedLocation, GRAB))
        return true
    }

    override fun handleMouseWheel(delta: Double): Boolean {
        return false
    }

    companion object {

        private fun playerRotToQuaternion(pitch: Double, yaw: Double): Quaterniond {
            return Quaterniond().rotateY(Math.toRadians(-yaw)).rotateX(Math.toRadians(pitch))
        }

        private fun dropShip(s: GravitronState, level: ServerLevel) {
            val grabbedShipId = s.shipID
            if (grabbedShipId != null) {
                val loadedShip = level.shipObjectWorld.loadedShips.getById(grabbedShipId)
                if (loadedShip != null) {
                    val gravitronForceInducer = getOrCreate(loadedShip)
                    gravitronForceInducer.data = null
                }
            }

            s.grabbing = false
            s.shipID = null
            s.shouldDrop = false
        }

        private fun updateShip(s: GravitronState, level: ServerLevel, entity: Entity) {
            if (s.grabbing) {
                val shipId = s.shipID
                if (shipId != null) {
                    val shipUnloaded = level.shipObjectWorld.allShips.getById(shipId)
                    val ship = level.shipObjectWorld.loadedShips.getById(shipId)
                    if (ship != null && s.playerGrabbedRotation != null && s.shipGrabbedDistance != null && s.shipGrabbedPos != null && s.heldBlockPos != null) {
                        // Update Rot Values
                        val playerCurrentRotation = Vector2d(entity.xRot.toDouble(), entity.yRot.toDouble())
                        val origPlayerRot =
                            playerRotToQuaternion(
                                s.playerGrabbedRotation!!.x(),
                                s.playerGrabbedRotation!!.y()
                            ).normalize()
                        val newPlayerRot =
                            playerRotToQuaternion(playerCurrentRotation.x(), playerCurrentRotation.y()).normalize()
                        val deltaPlayerRot = newPlayerRot.mul(origPlayerRot.conjugate(Quaterniond()), Quaterniond())
                        val rotation = deltaPlayerRot.mul(s.shipGrabbedRot, Quaterniond()).normalize()

                        // Update Pos Values
                        val lookDif = entity.lookAngle.toJOML().normalize().mul(s.shipGrabbedDistance!!)
                        s.heldBlockPos = entity.eyePosition.toJOML().add(lookDif)
                        val location = Vector3d(s.shipGrabbedPos)
                        val position = Vector3d(s.heldBlockPos)

                        val gravitronForceInducer = getOrCreate(ship)
                        val newData = GravitronForceInducerData(position, rotation, location)
                        gravitronForceInducer.data = newData
                    } else if (shipUnloaded == null) {
                        dropShip(s, level)
                    }
                }
            }
        }

        private fun updateShipDirection(s: GravitronState, level: ServerLevel, entity: Entity, dir: Direction) {
            if (s.grabbing) {
                val shipId = s.shipID
                if (shipId != null) {
                    val shipUnloaded = level.shipObjectWorld.allShips.getById(shipId)
                    val ship = level.shipObjectWorld.loadedShips.getById(shipId)
                    if (ship != null && s.playerGrabbedRotation != null && s.shipGrabbedDistance != null && s.shipGrabbedPos != null && s.heldBlockPos != null) {
                        // Update Rot Values
                        val lockedCurrentRotation = Vector2d(0.0, (dir.get2DDataValue() * 90).toDouble())
                        val origPlayerRot =
                            playerRotToQuaternion(
                                s.playerGrabbedRotation!!.x(),
                                s.playerGrabbedRotation!!.y()
                            ).normalize()
                        val newPlayerRot =
                            playerRotToQuaternion(lockedCurrentRotation.x(), lockedCurrentRotation.y()).normalize()
                        val deltaPlayerRot = newPlayerRot.mul(origPlayerRot.conjugate(Quaterniond()), Quaterniond())
                        val rotation = deltaPlayerRot.mul(s.shipGrabbedRot, Quaterniond()).normalize()

                        // Update Pos Values
                        val lookDif = entity.lookAngle.toJOML().normalize().mul(s.shipGrabbedDistance!!)
                        s.heldBlockPos = entity.eyePosition.toJOML().add(lookDif)
                        val location = Vector3d(s.shipGrabbedPos)
                        val position = Vector3d(s.heldBlockPos)

                        val gravitronForceInducer = getOrCreate(ship)
                        val newData = GravitronForceInducerData(position, rotation, location)
                        gravitronForceInducer.data = newData
                    } else if (shipUnloaded == null) {
                        dropShip(s, level)
                    }
                }
            }
        }

        @JvmStatic
        fun tick(player: Player) {
            if (player.level() is ServerLevel) {
                val s = getState(player)
                val graviton = player.getMainHandItem()
                val serverLevel = player.level() as ServerLevel

                if (!s.shouldDrop && graviton.`is`(ClockworkItems.GRAVITRON.asItem())) {
                    if (AllKeys.ACTIVATE_TOOL.isPressed) {
                        updateShipDirection(s, serverLevel, player, player.getDirection())
                    } else {
                        updateShip(s, serverLevel, player)
                    }
                } else {
                    dropShip(s, serverLevel)
                }

                if (s.grabCD != null && s.grabCD!! > 0) {
                    s.grabCD = s.grabCD!! - 1
                }
                if (graviton.hasTag() && graviton.tag!!.contains("GrabbedPosInShip") && !player.cooldowns.isOnCooldown(
                        graviton.item
                    )
                ) {
                    s.grabbing = true
                    val tag = graviton.tag

                    val clickLocation = readVec3(tag!!.getList("GrabbedPosInShip", Tag.TAG_DOUBLE.toInt()))
                    val id = tag.getLong("ShipId")

                    val ship: LoadedServerShip? = serverLevel.shipObjectWorld.loadedShips.getById(id)
                    if (ship != null) {
                        val transformedPos = ship.worldToShip.transformPosition(clickLocation.toJOML(), Vector3d())
                        grabShip(s, player, ship, transformedPos)
                        graviton.removeTagKey("ShipId")
                        graviton.removeTagKey("GrabbedPosInShip")
                    }
                }
            }
        }

        @JvmStatic
        fun tryGrabShip(level: ServerLevel, player: Player, clickedPos: BlockPos, clickLocation: Vec3): Boolean {
            val chunkX = clickedPos.x shr 4
            val chunkZ = clickedPos.z shr 4
            val ship = level.shipObjectWorld.loadedShips.getByChunkPos(chunkX, chunkZ, level.dimensionId)
            val grabPosInShip: Vector3dc = clickLocation.toJOML()
            val grabPosInWorld = Vector3d(grabPosInShip)
            val s = getState(player)
            if (level.isBlockInShipyard(clickedPos) && ship == null) {
                return false
            }

            if (ship == null) {
                return false
            } else {
                ship.shipToWorld.transformPosition(grabPosInWorld)
            }
            grabShip(s, player, ship, grabPosInShip)
            return true
        }

        private fun grabShip(s: GravitronState, p: Player, ship: LoadedServerShip, grabPosInShip: Vector3dc) {
            val heldPosInWorld = Vector3d()
            ship.transform.shipToWorld.transformPosition(Vector3d(grabPosInShip), heldPosInWorld)

            s.shipID = ship.id
            s.heldBlockPos = heldPosInWorld
            s.playerGrabbedRotation = Vector2d(p.xRot.toDouble(), p.yRot.toDouble())
            s.shipGrabbedPos = Vector3d(grabPosInShip)
            s.shipGrabbedRot = ship.transform.shipToWorldRotation
            s.shipGrabbedDistance = p.eyePosition.toJOML().distance(heldPosInWorld)
            ship.isStatic = false
        }
    }


}